x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStoreCacheTests.java

```diff
@@ -0,0 +1,303 @@
+import org.elasticsearch.xpack.core.security.action.user.HasPrivilegesResponse;
+import org.elasticsearch.xpack.core.security.action.user.PutUserRequestBuilder;
+import org.elasticsearch.xpack.core.security.action.user.PutUserResponse;
+import org.elasticsearch.xpack.core.security.authc.support.Hasher;
+import org.elasticsearch.xpack.core.security.authz.RoleDescriptor;
+import org.elasticsearch.xpack.core.security.authz.privilege.ApplicationPrivilegeDescriptor;
+import org.junit.Before;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+
+import static java.util.Collections.emptyMap;
+import static java.util.Collections.singleton;
+import static java.util.Collections.singletonMap;
+import static org.elasticsearch.action.support.WriteRequest.RefreshPolicy.IMMEDIATE;
+import static org.elasticsearch.test.SecuritySettingsSource.TEST_PASSWORD_HASHED;
+import static org.elasticsearch.test.SecuritySettingsSource.TEST_ROLE;
+import static org.elasticsearch.test.SecuritySettingsSourceField.TEST_PASSWORD;
+import static org.elasticsearch.xpack.core.security.authz.privilege.ApplicationPrivilegeDescriptor.DOC_TYPE_VALUE;
+import static org.elasticsearch.xpack.core.security.index.RestrictedIndicesNames.SECURITY_MAIN_ALIAS;
+
+public class NativePrivilegeStoreCacheTests extends SecuritySingleNodeTestCase {
+
+    private static final String APP_USER_NAME = "app_user";
+
+    @Override
+    protected String configUsers() {
+        return super.configUsers()
+            + APP_USER_NAME + ":" + TEST_PASSWORD_HASHED + "\n";
+    }
+
+    @Override
+    protected String configRoles() {
+        return super.configRoles()
+            + "app_role:\n"
+            + "  cluster: ['monitor']\n"
+            + "  indices:\n"
+            + "    - names: ['*']\n"
+            + "      privileges: ['read']\n"
+            + "  applications:\n"
+            + "    - application: 'app-1'\n"
+            + "      privileges: ['read', 'check']\n"
+            + "      resources: ['foo']\n"
+            + "    - application: 'app-2'\n"
+            + "      privileges: ['check']\n"
+            + "      resources: ['foo']\n";
+    }
+
+    @Override
+    protected String configUsersRoles() {
+        return super.configUsersRoles()
+            + "app_role:" + APP_USER_NAME + "\n"
+            + TEST_ROLE + ":" + APP_USER_NAME + "\n";
+    }
+
+    @Before
+    public void configureApplicationPrivileges() {
+        final List<ApplicationPrivilegeDescriptor> applicationPrivilegeDescriptors = Arrays.asList(
+            new ApplicationPrivilegeDescriptor("app-1", "read", Set.of("a:b:c", "x:y:z"), emptyMap()),
+            new ApplicationPrivilegeDescriptor("app-1", "write", Set.of("a:b:c", "x:y:z"), emptyMap()),
+            new ApplicationPrivilegeDescriptor("app-1", "admin", Set.of("a:b:c", "x:y:z"), emptyMap()),
```

Comment:
Should the action names be different? It seems hard to check that the right value is cached when they're all the same.